include/mpsc.h

```diff
@@ -128,8 +128,12 @@ class mpsc_queue_t
     {
       thread_id = my_thread_dbug_id();
       m_single_consumer_check = true;
-    } else
-      DBUG_ASSERT(thread_id == my_thread_dbug_id());
+    } else {
+      // turning off this assert as it might be necessary to
+      // dequeue from different thread and it
+      // seems my_thread_init is getting called in ways I don't understand
+      //DBUG_ASSERT(thread_id == my_thread_dbug_id());
```

Comment:
I'm concerned that this being commented out it implies that you have multiple consumer threads.  How do you know that there is never more than one consumer at a time?  A possible way would be something like this:

```
 class mpsc_queue_t
 {
 private:
#if !defined(DBUG_OFF)
  std::atomic<bool> dequeue_in_use;
#endif
...
  bool dequeue(T& output)
  {
    DBUG_ASSERT(!dequeue_in_use.exchange(true));  // Make sure it was false before

    buffer_node_t* tail = _tail.load(std::memory_order_relaxed);
    buffer_node_t* next = tail->next.load(std::memory_order_acquire);

    bool result;
    if (next == nullptr)
    {
      result = false;
    }
    else
    {
      output = next->data;
      _tail.store(next, std::memory_order_release);
      delete tail;
      result = true;
    }

    DBUG_ASSERT(dequeue_in_use.exchange(false));  // Make sure it was true before
    return result;
  }